Protocolli di comunicazione (#196)
Il-Libro-Open-Source/book: docs/it/protocolli-di-comunicazione.md

Questi infatti sono un insieme di regole che definiscono come avviene la comunicazione tra due o più entità, in questo caso tra due o più computer. Queste regole definiscono come avviene la comunicazione, come vengono trasmesse le informazioni e come vengono interpretate.

L'esempio più comune nel nostro ambito lavorativo è quello delle chiamate HTTP, che avvengono tra un client e un server, dove, per esempio, il client è il nostro browser e il server è il server web che ospita il sito che stiamo visitando. In questo caso il protocollo HTTP definisce come avviene la comunicazione tra il nostro browser e il server web, come vengono trasmesse le informazioni e come queste vengono interpretate.
Essendo così diffuso, spesso le specifiche del protocollo HTTP vengono date per scontate ma è importante tenerne conto, soprattutto quando si sviluppano applicazioni che sfruttano questo protocollo.

> Il protocollo HTTP è solo un esempio: esistono molti altri protocolli di comunicazione, e variano molto in base alla loro diffusione, complessita, sicurezza garantita, performance e contesto di utilizzo.

Reviewer: ```suggestion
Il protocollo HTTP è solo un esempio: esistono molti altri protocolli di comunicazione, e variano molto in base alla loro diffusione, complessità, sicurezza garantita, performance e contesto di utilizzo.
```

I protocolli testuali sono protocolli che utilizzano un formato testuale per trasmettere le informazioni, ovvero i bytes trasmessi hanno un corrispettivo nell'alfabeto, tipicamente tramite la codifica ASCII o UTF8.
Questo significa che le informazioni vengono trasmesse in un formato che è sia leggibile da un essere umano sia da un computer. Questo tipo di protocolli sono meno efficienti e veloci, ma sono più facili da implementare e da utilizzare. Inoltre, essendo leggibili da un essere umano, sono più facili da analizzare in caso di problemi. L'esempio più comune di protocollo testuale è il protocollo HTTP/1.1.

> I protocolli binari, invece, sono protocolli che utilizzano un formato binario per trasmettere le informazioni, ovveri i bytes trasmetti hanno un significato ben preciso, solitamente diverso dalla codifica ASCII. Questo significa che le informazioni vengono trasmesse in un formato che non è leggibile da un essere umano, ma che è leggibile da un computer. Questo tipo di protocolli sono più efficienti e veloci, ma sono più difficili da implementare e da utilizzare. Esempi di protocolli binari sono il protocollo TCP, UDP, HTTP2 e gRPC.

Reviewer: ```suggestion
I protocolli binari, invece, sono protocolli che utilizzano un formato binario per trasmettere le informazioni, ovvero i bytes trasmessi hanno un significato ben preciso, solitamente diverso dalla codifica ASCII. Questo significa che le informazioni vengono trasmesse in un formato che non è leggibile da un essere umano, ma che è leggibile da un computer. Questo tipo di protocolli sono più efficienti e veloci, ma sono più difficili da implementare e da utilizzare. Esempi di protocolli binari sono il protocollo TCP, UDP, HTTP2 e gRPC.
```

Entrambi i tipi di protocolli sono utilizzati e utilizzano lo stesso principio di base: definire un formato per trasmettere le informazioni e serializzare le informazioni in bytes. La differenza è che i protocolli testuali utilizzano una codifica testuale, mentre i protocolli binari utilizzano una codifica dedicata e binaria.

### Lo stack OSI e la distinzione tra TCP e UDP

> In base al modello [OSI](https://it.wikipedia.org/wiki/Modello_OSI), i protocolli di comunicazione possono essere divisi in vari livelli ma il concetto di base è il seguente: strutturare la comunicazione in più livelli, dove ogni livello ha un compito specifico e viene definito basandosi sui livelli precedenti. Per esempio, il protocollo HTTP/1.1 è definito al livello 7 e utilizza il protocollo TCP, che è definito invece al livello 4, il quale di basa sul protocollo IP, definito al livello 3. Questo permette di definire protocolli più complessi basandosi su protocolli più semplici e di riutilizzare i protocolli più semplici in più contesti.

Reviewer: ```suggestion
In base al modello [OSI](https://it.wikipedia.org/wiki/Modello_OSI), i protocolli di comunicazione possono essere divisi in vari livelli ma il concetto di base è il seguente: strutturare la comunicazione in più livelli, dove ogni livello ha un compito specifico e viene definito basandosi sui livelli precedenti. Per esempio, il protocollo HTTP/1.1 è definito al livello 7 e utilizza il protocollo TCP, che è definito invece al livello 4, il quale si basa sul protocollo IP, definito al livello 3. Questo permette di definire protocolli più complessi basandosi su protocolli più semplici e di riutilizzare i protocolli più semplici in più contesti.
```

Senza però entrare troppo nei dettagli, possiamo dire che i protocolli di comunicazione possono essere divisi in due categorie: quelli orientati alla connessione e quelli non orientati alla connessione.

Il protocollo TCP è un protocollo di comunicazione orientato alla connessione, questo significa che prima di trasmettere le informazioni, il client e il server devono stabilire una connessione tramite un handshake. Questo meccanismo permette di inizializzare la connessione che garantirà che le informazioni vengano trasmesse correttamente e che non vengano perse durante la trasmissione.

> Il protocollo UDP è un protocollo di comunicazione non orientato alla connessione, questo significa che non è necessario effettuare meccanismi di inizializzazione per la trasmissione delle informazioni. Vista la mancanza di azioni iniziali come quella dell'handshake, questo tipo di procolli permette di trasmettere le informazioni più velocemente, ma non ne garantisce la correttezza o che esse non vengano perse.

Author: ```suggestion
Il protocollo UDP è un protocollo di comunicazione non orientato alla connessione, questo significa che non è necessario effettuare meccanismi di inizializzazione per la trasmissione delle informazioni. Vista la mancanza di azioni iniziali come quella dell'handshake, questo tipo di protocolli permette di trasmettere le informazioni più velocemente, ma non ne garantisce la correttezza o che esse non vengano perse.
```

Il modello OSI è una struttura concettuale che ci permette di identificare e catalogare i protocolli di comunicazione, ma solitamente non viene utilizzato direttamente nella pratica. In pratica, i protocolli di comunicazione vengono utilizzati in modo trasparente, senza dover conoscere i dettagli del modello OSI.

### Il protocollo DNS per la risoluzione dei nomi di dominio

Durante la navigazione web, siamo soliti utilizzare i nomi di dominio per accedere ai siti web, come ad esempio `www.google.com`. Questi nomi di dominio sono più facili da ricordare, utilizzare e da condividere. Purtroppo però, i computer non possono utilizzare direttamente i nomi di dominio per connettersi ad un server web: è necessario convertire il nome di dominio in un indirizzo IP. Il protocollo DNS (Domain Name System) è utilizzato per rispondere a questa esigenza.

> Il client, prima di effettuare la richiesta ad un server web, deve quindi tradurre il nome di dominio in un indirizzo IP. Per fare questo, il client invia una richiesta ad un server DNS, un server che contiene un database la corrispondenza tra i nomi di dominio e loro indirizzi IP. Solo quando il client riceve dal server DNS l'indirizzo IP corrispondente al nome di dominio, può effettuare la richiesta al server web.

Reviewer 2: ```suggestion
Il client, prima di effettuare la richiesta ad un server web, deve quindi tradurre il nome di dominio in un indirizzo IP. Per fare questo, il client invia una richiesta ad un server DNS, ossia un server avente nel proprio database la corrispondenza tra i nomi di dominio e loro indirizzi IP. Solo quando il client riceve dal server DNS l'indirizzo IP corrispondente al nome di dominio, può effettuare la richiesta al server web.
```

Reviewer 2: @allevo Cosa ne pensi di questa suggestion? Pensi sia da integrare o preferisci scartarla? Perché nel primo caso la portiamo dentro e procediamo, nel secondo caso procediamo e basta 😀

Rimango in attesa di un tuo feedback 👋

Con la versione 1.1, invece, la connessione TCP viene mantenuta aperta e riutilizzata per effettuare più richieste sequenziali. Questo favorisce le prestazioni e la velocità di trasmissione delle informazioni.

Le versioni 1.x dell'HTTP avevano tutte un grosso limite: per effettuare 2 richieste in parallelo si dovevano utilizzare 2 connessioni TCP differenti. Con la versione 2 del protocollo HTTP questo limite viene superato grazie al concetto di multiplexing: il protocollo HTTP/2 utilizza infatti un unico canale di comunicazione TCP per trasmettere più richieste in parallelo. Questo permette di migliorare le prestazioni e la velocità di trasmissione delle informazioni. Questa modifica migliora nettamente le performance in linea generale, ma in particolar modo delle applicazioni web che utilizzano molte risorse esterne come ad esempio le immagini, i file CSS e i file JavaScript. La versione 2 utilizza un nuovo formato binario per trasmettere le informazioni.

> La versione 3 del protocollo HTTP viene introdotto per migliorare la sicurezza delle connessioni. Infatti con questa versione, la comunicazione è forzata ad essere criptata e migliora le tempistiche per la creazione di una connessione sicura.

Reviewer: ```suggestion
La versione 3 del protocollo HTTP viene introdotta per migliorare la sicurezza delle connessioni. Infatti con questa versione, la comunicazione è forzata ad essere criptata e migliora le tempistiche per la creazione di una connessione sicura.
```

> Un ulteriore cambio tecnologico introdotto è l'utilizzo del protocollo UDP: tale scelta è stata necessaria visti i problemi riscontrati con HTTP versione 2 per cui il multiplexing in casi di errore di trasmissione non era performante. Pur prevedendo infatti il multiplexing, l'invio dei dati avviene in modo sequenziale: la gestione dei errori di trasmissione del protocollo TCP cu sui la versione 2 si basa soffre del problema noto come "head-of-line".

Reviewer: ```suggestion
Un ulteriore cambio tecnologico introdotto è l'utilizzo del protocollo UDP: tale scelta è stata necessaria visti i problemi riscontrati con HTTP versione 2 per cui il multiplexing in casi di errore di trasmissione non era performante. Pur prevedendo infatti il multiplexing, l'invio dei dati avviene in modo sequenziale: la gestione degli errori di trasmissione del protocollo TCP cu sui la versione 2 si basa soffre del problema noto come "head-of-line".
```

> Se sulla connessione vengono inviati tre pacchetti, e solo il secondo ha un errori di trasmissione, il client è costretto a inviare nuovamente gli ultimi due pacchetti anche se l'ultimo ha raggiunto la destinazione correttamente. La versione 3 basata su UDP, non soffre di questo problema per cui è consigliata in modo particolare in casi di bassa connettività o dove il numero di errori di tramsissione è alto.

Reviewer 3: tramsissione -> trasmissione

> Se sulla connessione vengono inviati tre pacchetti, e solo il secondo ha un errori di trasmissione, il client è costretto a inviare nuovamente gli ultimi due pacchetti anche se l'ultimo ha raggiunto la destinazione correttamente. La versione 3 basata su UDP, non soffre di questo problema per cui è consigliata in modo particolare in casi di bassa connettività o dove il numero di errori di tramsissione è alto.

Reviewer 4: ```suggestion
Se sulla connessione vengono inviati tre pacchetti, e solo il secondo ha un errori di trasmissione, il client è costretto a inviare nuovamente gli ultimi due pacchetti anche se l'ultimo ha raggiunto la destinazione correttamente. La versione 3 basata su UDP, non soffre di questo problema per cui è consigliata in modo particolare in casi di bassa connettività o dove il numero di errori di tramissione è alto.
```

> Una menzione al protocollo HTTP (Hypertext Transfer Protocol Secure): non si tratta di un protocollo a tutti gli effetti, ma di un'estensione del protocollo HTTP, utile in particolare quando la connessione non è garantita sicura dal protocollo stesso, per esempio HTTP1.1 e HTTP2. In tali occasione, per rendere la comunicazione sicuro è necessario fare affidamento su un layer di criptazione.

Reviewer 2: ```suggestion
Una menzione al protocollo HTTPS (Hypertext Transfer Protocol Secure): non si tratta di un protocollo a tutti gli effetti, ma di un'estensione del protocollo HTTP, utile in particolare quando la connessione non è garantita sicura dal protocollo stesso, per esempio HTTP1.1 e HTTP2. In tali occasione, per rendere la comunicazione sicuro è necessario fare affidamento su un layer di criptazione.
```

> Una menzione al protocollo HTTPS (Hypertext Transfer Protocol Secure): non si tratta di un protocollo a tutti gli effetti, ma di un'estensione del protocollo HTTP, utile in particolare quando la connessione non è garantita sicura dal protocollo stesso, per esempio HTTP1.1 e HTTP2. In tali occasione, per rendere la comunicazione sicuro è necessario fare affidamento su un layer di criptazione.

Reviewer 4: ```suggestion
Una menzione va fatta anche al protocollo HTTPS (Hypertext Transfer Protocol Secure): non si tratta di un protocollo a tutti gli effetti, ma di un'estensione del protocollo HTTP, utile in particolare quando la connessione non è garantita sicura dal protocollo stesso, per esempio HTTP1.1 e HTTP2. In tali occasione, per rendere la comunicazione sicuro è necessario fare affidamento su un layer di criptazione.
```

> Non esiste un protocollo HTTP migliore di un altro: mentre è da sconsigliare i protocolli deprecati come HTTP 0.9 e HTTP1.0, gli altri rimangono molto validi.

Reviewer 2: ```suggestion
Non esiste un protocollo HTTP migliore di un altro: mentre è da sconsigliare i protocolli deprecati come HTTP 0.9 e HTTP 1.0, gli altri rimangono molto validi.
```

> Non esiste un protocollo HTTP migliore di un altro: mentre è da sconsigliare i protocolli deprecati come HTTP 0.9 e HTTP 1.0, gli altri rimangono molto validi.

Reviewer: ```suggestion
Non esiste un protocollo HTTP migliore di un altro: mentre sono da sconsigliare i protocolli deprecati come HTTP 0.9 e HTTP 1.0, gli altri rimangono molto validi.
```

> Per esempio per endpoint interni, invocati solo da applicazioni, il protocollo HTTP1.1 è solitamente la scelta utilizzata.

Reviewer 2: ```suggestion
Per esempio per endpoint interni, invocati solo da applicazioni, il protocollo HTTP 1.1 è solitamente la scelta utilizzata.
```

Il protocollo SSH (Secure Shell) è un protocollo di comunicazione basato su TCP e permette di stabilire una connessione sicura tra due computer.
Questo tipo di protocollo viene comunemente utilizzato per connettersi ad un computer remoto in modo sicuro, per trasferire file in modo sicuro e per eseguire comandi in modo sicuro. Il protocollo SSH è considerato ad oggi lo standard per eseguire queste operazioni.

Il protocollo è diviso in tre layer distinti:

- Transport Layer Protocol: è il layer che si occupa di stabilire una connessione sicura tra il client e il server. Qui vengono, per esempio, decisi gli algoritmi per la connessione. Vengono inoltre scambiate le chiavi pubbliche per garantire la sicurezza della connessione.
- User Authentication Protocol: è il layer che si occupa di autenticare l'utente che si sta connettendo al server. Vengono scambiate le credenziali dell'utente e vengono verificate dal server. Esempi di autenticazione sono la password e la chiave pubblica.
- Connection Layer Protocol: è il layer che si occupa di gestire i comandi del protocollo SSH, come ad esempio l'esecuzione di comandi remoti, il forwarding delle porte e il trasferimento dei file.

> Questa suddivisione permette di avere un protocollo flessibile e sicuro, che può essere configurato differentemente a seconda delle esigenze. Per esempio è possibile configurare il protocollo di compressione o di disabilitare l'autenticazione password, utilizzando solo la chiave pubblica.

Reviewer 4: ```suggestion
Questa suddivisione permette di avere un protocollo flessibile e sicuro, che può essere configurato differentemente a seconda delle esigenze. Per esempio è possibile configurare il protocollo di compressione o di disabilitare l'autenticazione tramite password, utilizzando solo la chiave pubblica.
```

La rappresentazione dei dati testuale è una rappresentazione che utilizza un formato testuale per trasmettere le informazioni. Questo significa che le informazioni vengono trasmesse in un formato che è leggibile anche da un essere umano, oltre che dal computer. Questo tipo di rappresentazione è meno efficiente e veloce, ma è più facile da implementare e da utilizzare. Inoltre, essendo leggibile da un essere umano, è più facile da analizzare in caso di problemi. Esempi più comuni di rappresentazione dei dati testuali sono il formato JSON, XML e YAML.

La rappresentazione dei dati binaria è una rappresentazione che utilizza un formato binario per trasmettere le informazioni. Questo significa che le informazioni vengono trasmesse in un formato che non è leggibile da un essere umano, ma che è leggibile da un computer. Questo tipo di rappresentazione è più efficiente e veloce, ma è più difficile da implementare e da utilizzare. Esempi più comuni di rappresentazione dei dati binari sono il formato Protocol Buffers, Avro e MessagePack.

> https://protobuf.dev/

Reviewer 4: ```suggestion

> Per approfondire:
https://protobuf.dev/
```

E inserirei anche gli altri due link nella nota

https://avro.apache.org/
https://msgpack.org/

### Il formato JSON

Il formato JSON (JavaScript Object Notation) è un formato testuale per la rappresentazione dei dati applicativi. Questo formato è basato sulla sintassi del linguaggio JavaScript e permette di rappresentare dati in modo semplice, strutturato e leggibile. Questo formato è molto diffuso e viene utilizzato in molti contesti, sia in ambito web sia in ambito mobile.

> Potremmo aver incontrato questo formato in diversi contesti, ad esempio quando si effettua una chiamata HTTP ad un server web, il server restituisce una risposta in formato JSON. Oppure in un file di configurazione.

Reviewer 4: ```suggestion
Potremmo aver incontrato questo formato in diversi contesti, ad esempio quando si effettua una chiamata HTTP ad un server web, il server restituisce una risposta in formato JSON, oppure in un file di configurazione.
```

Potremmo aver incontrato questo formato in diversi contesti, ad esempio in un servizio gRPC, essendo un formato binario e quindi più veloce e più efficiente di un formato testuale.

## Conclusioni

In questo capitolo abbiamo visto i principali protocolli di comunicazione, come funzionano e in quali ambiti vengono utilizzati. Abbiamo visto che i protocolli di comunicazione sono un insieme di regole che definiscono come avviene la comunicazione tra due o più entità, in questo caso tra due o più computer. Queste regole definiscono come avviene la comunicazione, come vengono trasmesse le informazioni e come vengono interpretate.
Non tutti i protocolli di comunicazione sono uguali: nel capitolo abbiamo dato solo alcuni esempi non esaustivi dei diversi utilizzi.

> Da notare inoltre che i protocolli non sono studiati a taovlino a priori ma rispecchiano l'evoluzione di diverse esigenze scaturite da problemi reali tecnologici, come per esempio migliorare le performance, la sicurezza o la gestione degli errori.

Reviewer 5: ```suggestion
Da notare inoltre che i protocolli non sono studiati a tavolino, ma rispecchiano l'evoluzione di diverse esigenze scaturite da problemi reali tecnologici, come per esempio migliorare le performance, la sicurezza o la gestione degli errori.
```

> Pur non essendo uno degli argomenti più gettonati e divertenti, spero che il lettore sia riuscito ad arrivare fino a qui in fondo!

Reviewer 4: ```suggestion
Pur non essendo uno degli argomenti più gettonati e divertenti, mi auguro che tu sia riuscito ad arrivare fino a qui in fondo e che l'argomento ti abbia stimolato un po' di curiosità!
```

---
layout: default
title: Protocolli di comunicazione
nav_order: 17
---

<!-- prettier-ignore-start -->
# Protocolli di comunicazione
{: .no_toc }

- TOC
{:toc}

<!-- prettier-ignore-end -->

#

> In ambito tecnologico e informatico sono presenti concetti fondamentali e strutture consolidate da tempo che spesso diamo per scontato. I protocolli di comunicazione sono un esempio di questa casistica: sono strumenti che utilizziamo quotidianamente, sia nella nostra vita lavorativa sia in quella privata. Ogni giorno infatti effettuiamo chiamate verso un server, inviamo dati, riceviamo risposte e tutto questo avviene grazie ai protocolli di comunicazione anche se, a volte, non ne siamo consapevoli.

Reviewer 6: ```suggestion
In ambito tecnologico e informatico esistono concetti fondamentali e strutture consolidate da tempo che spesso diamo per scontato. I protocolli di comunicazione sono un esempio di questa casistica: sono strumenti che utilizziamo quotidianamente, sia nella nostra vita lavorativa sia in quella privata. Ogni giorno infatti interroghiamo un server, inviamo dati, riceviamo risposte e tutto questo avviene grazie ai protocolli di comunicazione anche se, a volte, non ne siamo consapevoli.
```

Questi infatti sono un insieme di regole che definiscono come avviene la comunicazione tra due o più entità, in questo caso tra due o più computer e come le informazioni vengono trasmesse e interpretate.

> L'esempio più comune nel nostro ambito lavorativo è quello delle chiamate HTTP, che avvengono tra un client e un server, dove, per esempio, il client è il nostro browser e il server è il server web che ospita il sito che stiamo visitando. In questo caso il protocollo HTTP definisce come avviene la comunicazione tra il nostro browser e il server web, come vengono trasmesse le informazioni e come queste vengono interpretate.

Reviewer 6: Per evitare la ripetizione

```suggestion
L'esempio più comune nel nostro ambito lavorativo è quello delle chiamate HTTP, che avvengono tra un client e un server, dove, per esempio, il client è il nostro browser e il server è un computer che ospita il sito che stiamo visitando. In questo caso il protocollo HTTP definisce come avviene la comunicazione tra il nostro browser e il server web, come vengono trasmesse le informazioni e come queste vengono interpretate.
```

I protocolli binari, invece, sono protocolli che utilizzano un formato binario per trasmettere le informazioni, ovvero i bytes trasmessi hanno un significato ben preciso, solitamente diverso dalla codifica ASCII. Questo significa che le informazioni vengono trasmesse in un formato che non è leggibile da un essere umano, ma che è leggibile da un computer. Questo tipo di protocolli sono più efficienti e veloci, ma sono più difficili da implementare e da utilizzare. Esempi di protocolli binari sono il protocollo TCP, UDP, HTTP2 e gRPC.

Entrambi i tipi di protocolli sono utilizzati e utilizzano lo stesso principio di base: definire un formato per trasmettere le informazioni e serializzare le informazioni in bytes. La differenza è che i protocolli testuali utilizzano una codifica testuale, mentre i protocolli binari utilizzano una codifica dedicata e binaria.

### Lo stack OSI e la distinzione tra TCP e UDP

> In base al modello [OSI](https://it.wikipedia.org/wiki/Modello_OSI), i protocolli di comunicazione possono essere divisi in vari livelli ma il concetto di base è il seguente: strutturare la comunicazione in più livelli, dove ogni livello ha un compito specifico e viene definito basandosi sui livelli precedenti. Per esempio, il protocollo HTTP/1.1 è definito al livello 7 e utilizza il protocollo TCP, che è definito invece al livello 4, il quale si basa sul protocollo IP, definito al livello 3. Questo permette di definire protocolli più complessi basandosi su protocolli più semplici e di riutilizzare i protocolli più semplici in più contesti.

Reviewer 6: Mi suonava un pelo complicato da comprendere

```suggestion
In base al modello [OSI](https://it.wikipedia.org/wiki/Modello_OSI), i protocolli di comunicazione si dividono in vari livelli ma il concetto di base è il seguente: la comunicazione è strutturata su più livelli, ogni livello ha un compito specifico e viene definito basandosi sui livelli precedenti. Per esempio, il protocollo HTTP/1.1 è definito al livello 7 e utilizza il protocollo TCP, che è definito invece al livello 4, il quale si basa sul protocollo IP, definito al livello 3. Questo permette di definire protocolli più complessi basandosi su protocolli più semplici e di riutilizzare i protocolli più semplici in più contesti.
```

Questa tecnica è per esempio utilizzata quando si hanno dei nomi di dominio non pubblici, come ad esempio quelli utilizzati in una rete locale. In questo caso, il nome di dominio può essere risolto solo all'interno della rete locale, e non all'esterno, garantendo la sicurezza e la privacy della rete. Esempi tipici sono i nomi di dominio utilizzati in una rete aziendale come i nome di dominio di un database o la intranet aziendale.

Il meccanismo di traduzione viene usato in realtà per tutti i protocolli di comunicazione che richiedono la traduzione tra nome di dominio e indirizzo IP, non solo per il protocollo HTTP. Ad esempio, quando si invia una mail, il client deve tradurre il nome di dominio del server SMTP in un indirizzo IP per poter inviare la mail. Oppure, quando il backend di un'applicazione deve connettersi al database, deve tradurre il nome di dominio del database in un indirizzo IP per poter effettuare la connessione.

> Il protocollo DNS è complesso e supporta altre funzionalità oltre a quelle descritte sopra che putroppo per motivi pratici non sono coperte in questo capitolo.

Reviewer 3: una svista... "putroppo"

> Il protocollo DNS è complesso e supporta altre funzionalità oltre a quelle descritte sopra che putroppo per motivi pratici non sono coperte in questo capitolo.

Author: ```suggestion
Il protocollo DNS è complesso e supporta altre funzionalità oltre a quelle descritte sopra che purtroppo per motivi pratici non sono coperte in questo capitolo.
```

> Il protocollo DNS è complesso e supporta altre funzionalità oltre a quelle descritte sopra che putroppo per motivi di spazio non sono coperte in questo capitolo.

Reviewer 6: ```suggestion
Il protocollo DNS è complesso e supporta altre funzionalità oltre a quelle descritte sopra che putroppo per motivi pratici non sono coperte in questo capitolo.
```

Con la versione 1.1, invece, la connessione TCP viene mantenuta aperta e riutilizzata per effettuare più richieste sequenziali. Questo favorisce le prestazioni e la velocità di trasmissione delle informazioni.

Le versioni 1.x dell'HTTP avevano tutte un grosso limite: per effettuare 2 richieste in parallelo si dovevano utilizzare 2 connessioni TCP differenti. Con la versione 2 del protocollo HTTP questo limite viene superato grazie al concetto di multiplexing: il protocollo HTTP/2 utilizza infatti un unico canale di comunicazione TCP per trasmettere più richieste in parallelo. Questo permette di migliorare le prestazioni e la velocità di trasmissione delle informazioni. Questa modifica migliora nettamente le performance in linea generale, ma in particolar modo delle applicazioni web che utilizzano molte risorse esterne come ad esempio le immagini, i file CSS e i file JavaScript. La versione 2 utilizza un nuovo formato binario per trasmettere le informazioni.

> La versione 3 del protocollo HTTP viene introdotta per migliorare la sicurezza delle connessioni. Infatti con questa versione, la comunicazione è forzata ad essere criptata e migliora le tempistiche per la creazione di una connessione sicura.

Reviewer 6: ```suggestion
La versione 3 del protocollo HTTP è stata introdotta per migliorare la sicurezza delle connessioni. Infatti con questa versione, la comunicazione è forzata ad essere criptata e migliora le tempistiche per la creazione di una connessione sicura.
```

> Un ulteriore cambio tecnologico introdotto è l'utilizzo del protocollo UDP: tale scelta è stata necessaria visti i problemi riscontrati con HTTP versione 2 per cui il multiplexing in casi di errore di trasmissione non era performante. Pur prevedendo infatti il multiplexing, l'invio dei dati avviene in modo sequenziale: la gestione degli errori di trasmissione del protocollo TCP cu sui la versione 2 si basa soffre del problema noto come "head-of-line".

Reviewer 6: ```suggestion
Un ulteriore cambio tecnologico introdotto prevede l'utilizzo del protocollo UDP: tale scelta è stata necessaria visti i problemi riscontrati con HTTP versione 2 per cui il multiplexing in casi di errore di trasmissione non era performante. Pur prevedendo infatti il multiplexing, l'invio dei dati avviene in modo sequenziale: la gestione degli errori di trasmissione del protocollo TCP cu sui la versione 2 si basa soffre del problema noto come "head-of-line".
```

> Se sulla connessione vengono inviati tre pacchetti, e solo il secondo ha un errori di trasmissione, il client è costretto a inviare nuovamente gli ultimi due pacchetti anche se l'ultimo ha raggiunto la destinazione correttamente. La versione 3 basata su UDP, non soffre di questo problema per cui è consigliata in modo particolare in casi di bassa connettività o dove il numero di errori di tramissione è alto.

Reviewer: ```suggestion
Se sulla connessione vengono inviati tre pacchetti, e solo il secondo ha un errori di trasmissione, il client è costretto a inviare nuovamente gli ultimi due pacchetti anche se l'ultimo ha raggiunto la destinazione correttamente. La versione 3 basata su UDP, non soffre di questo problema per cui è consigliata in modo particolare in casi di bassa connettività o dove il numero di errori di trasmissione è alto.
```

> Se sulla connessione vengono inviati tre pacchetti, e solo il secondo ha un errori di trasmissione, il client è costretto a inviare nuovamente gli ultimi due pacchetti anche se l'ultimo ha raggiunto la destinazione correttamente. La versione 3 basata su UDP, non soffre di questo problema per cui è consigliata in modo particolare in casi di bassa connettività o dove il numero di errori di trasmissione è alto.

Reviewer 6: Typo

```suggestion
Se sulla connessione vengono inviati tre pacchetti, e solo il secondo ha un errore di trasmissione, il client è costretto a inviare nuovamente gli ultimi due pacchetti anche se l'ultimo ha raggiunto la destinazione correttamente. La versione 3 basata su UDP, non soffre di questo problema per cui è consigliata in modo particolare in casi di bassa connettività o dove il numero di errori di trasmissione è alto.
```

Una menzione va fatta anche al protocollo HTTPS (Hypertext Transfer Protocol Secure): non si tratta di un protocollo a tutti gli effetti, ma di un'estensione del protocollo HTTP, utile in particolare quando la connessione non è garantita sicura dal protocollo stesso, per esempio HTTP1.1 e HTTP2. In tali occasione, per rendere la comunicazione sicuro è necessario fare affidamento su un layer di criptazione.

Non esiste un protocollo HTTP migliore di un altro: mentre sono da sconsigliare i protocolli deprecati come HTTP 0.9 e HTTP 1.0, gli altri rimangono molto validi.
Per esempio per endpoint interni, invocati solo da applicazioni, il protocollo HTTP 1.1 è solitamente la scelta utilizzata.

### Il protocollo SSH

Il protocollo SSH (Secure Shell) è un protocollo di comunicazione basato su TCP e permette di stabilire una connessione sicura tra due computer.

> Questo tipo di protocollo viene comunemente utilizzato per connettersi ad un computer remoto in modo sicuro, per trasferire file in modo sicuro e per eseguire comandi in modo sicuro. Il protocollo SSH è considerato ad oggi lo standard per eseguire queste operazioni.

Reviewer 6: Per evitare la ripetizione

```suggestion
Questo tipo di protocollo viene comunemente utilizzato per connettersi ad un computer remoto, per trasferire file e per eseguire comandi, il tutto in modo sicuro. Il protocollo SSH è considerato ad oggi lo standard per eseguire queste operazioni.
```

La rappresentazione dei dati testuale è una rappresentazione che utilizza un formato testuale per trasmettere le informazioni. Questo significa che le informazioni vengono trasmesse in un formato che è leggibile anche da un essere umano, oltre che dal computer. Questo tipo di rappresentazione è meno efficiente e veloce, ma è più facile da implementare e da utilizzare. Inoltre, essendo leggibile da un essere umano, è più facile da analizzare in caso di problemi. Esempi più comuni di rappresentazione dei dati testuali sono il formato JSON, XML e YAML.

La rappresentazione dei dati binaria è una rappresentazione che utilizza un formato binario per trasmettere le informazioni. Questo significa che le informazioni vengono trasmesse in un formato che non è leggibile da un essere umano, ma che è leggibile da un computer. Questo tipo di rappresentazione è più efficiente e veloce, ma è più difficile da implementare e da utilizzare. Esempi più comuni di rappresentazione dei dati binari sono il formato Protocol Buffers, Avro e MessagePack.

Per approfondire:

- https://protobuf.dev/
- https://avro.apache.org/
- https://msgpack.org/

> ### Il formato JSON

Author: I titoli `JSON`, `XML`, `YAML`, `ProtoBuf` sono tutti H3 e stanno già sotto il cappello `La rappresentazione dei dati applicativi` che è un H2.
Se vuoi possiamo rivedere il titolo per essere meno legato allo stack ISO/OSI.

> ### Il formato JSON

Reviewer 4: No, direi che rileggendo va bene così. Grazie!

Potremmo aver incontrato questo formato in diversi contesti, ad esempio in un servizio gRPC, essendo un formato binario e quindi più veloce e più efficiente di un formato testuale.

## Conclusioni

In questo capitolo abbiamo visto i principali protocolli di comunicazione, come funzionano e in quali ambiti vengono utilizzati. Abbiamo visto che i protocolli di comunicazione sono un insieme di regole che definiscono come avviene la comunicazione tra due o più entità, in questo caso tra due o più computer. Queste regole definiscono come avviene la comunicazione, come vengono trasmesse le informazioni e come vengono interpretate.
Non tutti i protocolli di comunicazione sono uguali: nel capitolo abbiamo dato solo alcuni esempi non esaustivi dei diversi utilizzi.

> Da notare inoltre che i protocolli non sono studiati a tavolino, ma rispecchiano l'evoluzione di diverse esigenze scaturite da problemi reali tecnologici, come per esempio migliorare le performance, la sicurezza o la gestione degli errori.

Reviewer 4: Questa non è una nota, ma un commento per dire che questa è una frase pazzesca da usare anche per presentare il capitolo!

In ambito tecnologico e informatico sono presenti concetti fondamentali e strutture consolidate da tempo che spesso diamo per scontato. I protocolli di comunicazione sono un esempio di questa casistica: sono strumenti che utilizziamo quotidianamente, sia nella nostra vita lavorativa sia in quella privata. Ogni giorno infatti effettuiamo chiamate verso un server, inviamo dati, riceviamo risposte e tutto questo avviene grazie ai protocolli di comunicazione, anche se non ne siamo consapevoli.

Questi infatti sono un insieme di regole che definiscono come avviene la comunicazione tra due o più entità, in questo caso tra due o più computer. Queste regole definiscono come avviene la comunicazione, come vengono trasmesse le informazioni e come vengono interpretate.

L'esempio più comune nel nostro ambito lavorativo è quello delle chiamate HTTP, che avvengono tra un client e un server, dove il client è il nostro browser e il server è il server web che ospita il sito che stiamo visitando. In questo caso il protocollo HTTP definisce come avviene la comunicazione tra il nostro browser e il server web, come vengono trasmesse le informazioni e come queste vengono interpretate.

> Essendo così diffuso, spesso le speficiche del protocollo HTTP vengono date per scontate ma è importante tenerne conto, soprattutto quando si sviluppano applicazioni che sfruttano questo protocollo.

Reviewer 2: ```suggestion
Essendo così diffuso, spesso le specifiche del protocollo HTTP vengono date per scontate ma è importante tenerne conto, soprattutto quando si sviluppano applicazioni che sfruttano questo protocollo.
```

> Il protocollo HTTP è solo un esempio: esistono molti altri protocolli di comunicazione, alcuni più diffusi di altri, alcuni più complessi di altri, alcuni più sicuri di altri, alcuni più performanti di altri, alcuni più adatti di altri a determinati scopi.

Reviewer 2: @allevo D'accordissimo sul concetto, però mi sembra un po' troppo forzata la continua ripetizione di _"più `<qualcosa>` di altri"_...non trovi?

Senza però entrare troppo nei dettagli, possiamo dire che i protocolli di comunicazione possono essere divisi in due categorie: quelli orientati alla connessione e quelli non orientati alla connessione.

Il protocollo TCP è un protocollo di comunicazione orientato alla connessione, questo significa che prima di trasmettere le informazioni, il client e il server devono stabilire una connessione tramite un handshake. Questo meccanismo permette di inizializzare la connessione che garantirà che le informazioni vengano trasmesse correttamente e che non vengano perse durante la trasmissione.

> Il protocollo UDP è un protocollo di comunicazione non orientato alla connessione, questo significa che non è necessario effettuare maccanismi di inizializzazione per la trasmissione delle informazioni. Vista la mancanza di azioni come quella dell'handshake, questo tipo di procolli permette di trasmettere le informazioni più velocemente, ma non ne garantisce la correttezza o che esse non vengano perse.

Reviewer 3: procolli -> protocolli

Author: grazie!

Durante la navigazione di un sito web, il client deve inviare una richiesta ad un server web, che è identificato da un indirizzo IP. Questo indirizzo IP è un indirizzo numerico che identifica univocamente un computer. Poiché è difficile ricordare un indirizzo IP, è stato introdotto il concetto di nome di dominio che è un nome testuale che identifica univocamente un computer. Questo nome di dominio è più facile da ricordare e da utilizzare, ma non può essere utilizzato per connettersi direttamente ad un computer: sarà necessario infatti convertire il nome in indirizzo IP. Il protocollo DNS è utilizzato per fare questo.

Il client, prima di effettuare la richiesta ad un server web, deve tradurre il nome di dominio in un indirizzo IP. Per fare questo, il client invia una richiesta ad un server DNS, che è un server che contiene un database con tutti i nomi di dominio e gli indirizzi IP corrispondenti. Il server DNS, dopo aver ricevuto la richiesta, cerca il nome di dominio nel database e, se lo trova, restituisce l'indirizzo IP corrispondente. Il client, dopo aver ricevuto l'indirizzo IP, può effettuare la richiesta al server web.

> Questo meccanismo di tradzione viene usato in realtà per tutti i protocolli di comunicazione, non solo per il protocollo HTTP. Ad esempio, quando si invia una mail, il client deve tradurre il nome di dominio del server SMTP in un indirizzo IP per poter inviare la mail. Oppure, quando il backend di un'applicazione deve connettersi al database, deve tradurre il nome di dominio del database in un indirizzo IP per poter effettuare la connessione.

Reviewer 2: ```suggestion
Questo meccanismo di traduzione viene usato in realtà per tutti i protocolli di comunicazione, non solo per il protocollo HTTP. Ad esempio, quando si invia una mail, il client deve tradurre il nome di dominio del server SMTP in un indirizzo IP per poter inviare la mail. Oppure, quando il backend di un'applicazione deve connettersi al database, deve tradurre il nome di dominio del database in un indirizzo IP per poter effettuare la connessione.
```

Il protocollo DNS è complesso e supporta altre funzionalità che putroppo per motivi di spazio non sono coperte in questo capitolo.

https://it.wikipedia.org/wiki/Domain_Name_System

### Il protocollo HTTP

Il protocollo HTTP (HyperText Transfer Protocol) è sicuramente uno dei protocolli di comunicazione più diffusi nel web: agli inizi del web, era l'unico protocollo utilizzato e ancora oggi è il protocollo più utilizzato per trasmettere informazioni tra un client e un server. Questo protocollo è utilizzato per trasmettere informazioni tra un client e un server, dove il client è l'entità che vuole effettuare la richiesta e il server è l'entità che deve produrre la risposta.
Agli inizi del web, il protocollo HTTP era utilizzato per trasmettere informazioni testuali, come ad esempio il contenuto di una pagina HTML. Con il tempo, il protocollo HTTP è stato esteso per trasmettere anche altri tipi di informazioni, come ad esempio file multimediali, file binari, JSON, ecc.

> Nel corso degli anni, il protocollo HTTP ha avuto diverse revisioni per migliorare alcune caratteristiche come le prestazioni e la sicurezza. Infatti la prima versione è la version 0.9, oramai non più usata, che è stata utilizzata per trasmettere informazioni testuali descritta qui https://www.w3.org/Protocols/HTTP/AsImplemented.html.

Reviewer 2: ```suggestion
Nel corso degli anni, il protocollo HTTP ha avuto diverse revisioni atte a migliorarne alcune caratteristiche come le prestazioni e la sicurezza. Infatti la prima versione è la version 0.9 (oramai non più usata) che è stata utilizzata per trasmettere informazioni testuali, descritta qui: https://www.w3.org/Protocols/HTTP/AsImplemented.html.
```

> La versione 1.0 è la versione che ha introdotto il concetto di header e che ha permesso di trasmettere informazioni di contesto della chiamata come il `content-type` e i `cookie` ma è stata deprecata in favore della versione 1.1 che ha introdotto il concetto di connessione TCP persistente: permette infatti di mantenere la connessione aperta tra il client e il server per poterla rifruttare nuovamente per una nuova richiesta HTTP. Questo favorisce le prestazioni e la velocità di trasmissione delle informazioni.

Reviewer 2: ```suggestion
La versione 1.0 è la versione che ha introdotto il concetto di header e che ha permesso di trasmettere informazioni di contesto della chiamata come il `content-type` e i `cookie` ma è stata deprecata in favore della versione 1.1 che ha introdotto il concetto di connessione TCP persistente: essa permette infatti di mantenere la connessione aperta tra il client e il server per poterla sfruttare nuovamente per una nuova richiesta HTTP. Questo favorisce le prestazioni e la velocità di trasmissione delle informazioni.
```

> Le versioni 1.x dell'HTTP avevano tutte un grosso limite: per effettuare 2 richieste in parallelo si doveva utilizzare 2 connessioni TCP differenti. Con la versione 2 del protocollo HTTP questo limite viene superato grazie ad un formato binario che permette di trasmettere più informazioni contemporaneamente sulla stessa connessione. Questa modifica migliora nettamente le performance in modo particolare per le applicazioni web che utilizzano molte risorse esterne come ad esempio le immagini, i file CSS e i file JavaScript.

Reviewer 2: ```suggestion
Le versioni 1.x dell'HTTP avevano tutte un grosso limite: per effettuare 2 richieste in parallelo si dovevano utilizzare 2 connessioni TCP differenti. Con la versione 2 del protocollo HTTP questo limite viene superato grazie ad un formato binario che permette di trasmettere più informazioni contemporaneamente sulla stessa connessione. Questa modifica migliora nettamente le performance in linea generale, ma in particolar modo delle applicazioni web che utilizzano molte risorse esterne come ad esempio le immagini, i file CSS e i file JavaScript.
```

> La versione 3 del protocollo HTTP cambia lo stack OSI precedentemente basato su TCP: utilizzando un protocollo UDP, questa versione forza l'uso di una connessione cryptata e riduce l'inizializzazione della connessione. Pur essendo basata su un protocollo UDP, questa versione mantiene le stesse caratteristiche delle versioni precedenti, come ad esempio il concetto di header e di richiesta e risposta, il formato binario ottimizzando le prestazioni su connessioni lente.

Reviewer 2: ```suggestion
La versione 3 del protocollo HTTP cambia lo stack OSI precedentemente basato su TCP: utilizzando un protocollo UDP, questa versione forza l'uso di una connessione cryptata e riduce l'inizializzazione della connessione. Pur essendo basata su un protocollo UDP, questa versione mantiene le stesse caratteristiche delle versioni precedenti, come ad esempio il concetto di header, di richiesta e risposta e del formato binario, ottimizzando le prestazioni su connessioni lente.
```

### Il formato Protocol Buffers

Il formato Protocol Buffers è un formato binario per la rappresentazione dei dati applicativi. Questo formato è basato sulla sintassi del linguaggio Protocol Buffers e permette di rappresentare dati in modo strutturato e veloce. Questo formato è molto diffuso poiché permette di definire un formato personalizzato per la rappresentazione dei dati, usando una definizione di schema. Permette di gestire in modo semplice e veloce la serializzazione e la deserializzazione dei dati anche in modo retrocompatibile.
Potremmo aver incontrato questo formato in diversi contesti, ad esempio in un servizio gRPC, essendo un formato binario e quindi più veloce e più efficiente di un formato testuale.

## Conclusioni

In questo capitolo abbiamo visto i principali protocolli di comunicazione, come funzionano e in quali ambiti vengono utilizzati. Abbiamo visto che i protocolli di comunicazione sono un insieme di regole che definiscono come avviene la comunicazione tra due o più entità, in questo caso tra due o più computer. Queste regole definiscono come avviene la comunicazione, come vengono trasmesse le informazioni e come vengono interpretate.

> Non tutti i protocolli di comunicazione sono uguali: alcuni sono più diffusi di altri, alcuni più complessi di altri, alcuni più sicuri di altri, alcuni più performanti di altri, alcuni più adatti di altri a determinati scopi. È importante quindi scegliere il protocollo di comunicazione più adatto al contesto in cui si lavora.

Reviewer 2: @allevo Qui ho il medesimo dubbio di [qui](https://github.com/Il-Libro-Open-Source/book/pull/196/files#diff-3fe0312258659ab0d7aa36e42decef10d95bf10d9632d334c2fc610e2d142517R10), questa continua ripetizione, durante la lettura della frase, ha qualcosa che non mi convince appieno.

Fermo restando che qualora fosse grammaticalmente corretta e risultasse buona in fase di lettura, la troverei comunque una ripetizione della frase di cui sopra.
